### ISSUE TITLE
Filecache garbage collector & many parallel requests generated warnings

### ISSUE THREAD

**Issue by pmacko** [cakephp/cakephp#12515]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.6.10
* Platform and Target: Nginx, php-fpm, MySQL

### What happened
Filecache garbage collector & many parallel requests generater warnings like this
E_WARNING: SplFileObject::__construct(.../tmp/cache/15minutes/15minutes_xyz_690): failed to open stream: No such file or directory

Garbage collection performs more parallel processes at a time (in one second) and therefore this error occurs

### Solution

file: /src/Cache/CacheRegistry.php, line: 96
https://github.com/cakephp/cakephp/blob/master/src/Cache/CacheRegistry.php#L96
```
$config = $instance->getConfig();
if ($config['probability'] && time() % $config['probability'] === 0) {
    $instance->gc();
}
```
replace with this:
```
$config = $instance->getConfig();
$random = function_exists('random_int') ? 'random_int' : 'mt_rand';
if ($config['probability'] && $random(0, $config['probability']) === 0) {
    $instance->gc();
}
```

**Comment by lorenzo** [comment]
I think the learning here is to not use FileCache under a highly concurrent load pattern. Better use App, or Redis :)

**Comment by pmacko** [comment]
Yes I agree. But there are situations where it is not possible. Already at 100 req / min. this sometimes happens.
However, the probability should not be counted as `time() % x` :)

**Comment by lorenzo** [comment]
You can disable the GC, there is no good way to solve this other than not using the File cache under those situations. It was not built with those cases in mind.

**Comment by sukihub** [comment]
@lorenzo you're absolutely right, no arguments there :)

Still, emptying cache only if request comes exactly on 100th second doesn't seem that good even for low traffic environments (small sites or development machines). What is the probability of request actually hitting that?

So no. The change will not help high load environments. But it just seems like more reasonable default, even if only for development. And it's low effort - the solution is already there :)

**Comment by sukihub** [comment]
Plus, as @pmacko just pointed out to me, the code actually affects all caches that (possibly) use GC - so it's not just file cache.

The other option would be removing GC mechanism completely, since it's not really being used and is kind of broken anyway :)

**Comment by ADmad** [comment]
No "real" cache engine besides FileEngine actually does gc. So it would be best to remove the method in 4.x.

**Comment by lorenzo** [comment]
We could just remove GC, it is only ever used for avoiding having too many files in the file system while developing.
